### DISCUSSION TITLE
Managing buffer(s) for dynamic meshes

### DISCUSSION THREAD

**Discussion by 0x241F31** [gpuweb/gpuweb#3411] [Q&A]
I'm trying to implement drawing with WebGPU, so dealing with growing mesh. Also, I'm keeping in mind that it could be collaborative, so it could be multiple growing meshes. 
I'm searching for correct approach here, with minimal allocating / drawcalls.

As you can see, it overflows buffer at some point:
https://user-images.githubusercontent.com/17229619/188286311-382d8955-ae39-4113-8fd5-6567f57eceac.gif

First time I made it obviously simple: every frame / user input I create new buffer and populate it with new calculated VBO/IBO. However, I thought that it's not good for performance. Is that right?

Then I thought I can allocate big chunks for every drawing that is currently being drawed by user(s): `|Mesh1_____|Mesh2____|LargeMesh3_|`. And at the end of drawing place it to another static buffer. So it would be 2 draw calls. But, after implementation of this, I faced fact that only first chunk is being rendered. As I understood, this is due that shader see empty space (___) as vertices and indices with zero value. That's how I'm doing that: 
```ts
alloc_item() {
   const newVBO = new VBO_ARRAY(mem.VBO.length + VBO_CHUNK_LENGTH)
   newVBO.set(mem.VBO)
   mem.VBO = newVBO
   mem.vboBuffer = createBufferFromArray(device, mem.VBO, VBO_USAGE)
   // ibo etc...
}
growMesh() {
   // compute new mesh....
    mem.VBO.set(mesh.vertices, vOffset)
    device.queue.writeBuffer(
       mem.vboBuffer,
       vOffset * VBO_ARRAY.BYTES_PER_ELEMENT,
       mem.VBO.slice(vOffset, vOffset + vLength)
    )
}
```
Is it expected behaviour?  

Overall, what's good approach here? Is using writeBuffer correct? 
I'm new in CG so would appreciate any advices on this topic. 

BTW, thank you for your wok on this technology!

**Comment by ghadeeras** [comment]
Hello!

A little disclaimer first: I am not one of the contributors to these new specs, and I am no CG expert :-) but I hope I could provide some pointers.

In my modest experience, for dynamic volatile buffers, specially when it is difficult to put a cap on the size of the buffer, I often opted for the simple solution of destroying old buffers with old meshes, and creating new buffers with new meshes that get consumed by render passes. Care must be taken not to submit a destroyed buffer to a render pass.

I do hope more experienced passersby would keep me honest and correct me if I am wrong.

In any way, I think you might find the practices recommended in this link insightful: https://toji.github.io/webgpu-best-practices/ ... They're written by one of the contributors and there's a specific page there for buffers.

Cheers!

**Reply by 0x241F31** [reply]
Thanks! I'll try to benchmark this

**Comment by kainino0x** [comment]
Another approach would be to create a new buffer and issue multiple draw calls - one for each chunk of the mesh.

**Reply by ghadeeras** [reply]
Right, this is even better than my suggestion, given the mesh is being augmented incrementally (if I understood correctly).

**Reply by 0x241F31** [reply]
Thank you. In principle, the easiest option is to make a separate drawcall for each mesh, but I am worried that with a large number of meshes (in my case, hand-drawn lines or figures) there will be a performance problem.
I would like to meet 120 fps on mobile devices.

**Reply by kainino0x** [reply]
If you get to a point where draw calls are a limiting factor (probably 1k-10k draw calls at least), then your buffer must also be quite large, and the cost of copying the buffer into a new, larger buffer just to resize it would most likely be much worse. Ideas:

- Once a stroke is done, consolidate all of its buffers into one so it becomes just one draw call
- Similar, but consolidate even more than one stroke at a time for parts of the drawing that are relatively static over time - this will require a "caching" heuristic
- Reduce the number of calls needed by grouping draw calls into render bundles, which are reusable. Will also require a caching heuristic, but avoids moving a lot of data around between buffers

**Reply by 0x241F31** [reply]
Yep, I figured out term "Draw call batching" is suitable for this case. However, now I'm challinging problem: my strokes are single colored, so I want to reduce VBO size and send color per stroke, not per vertex, I was able to do it with multiple draw calls using second vertex buffer for colors with `stepMode: 'instance'`, however, I have no idea how to achive it in case of batched single draw call. I asked on SO: https://stackoverflow.com/questions/73638228/set-color-for-the-whole-mesh-but-not-every-vertex

**Reply by kainino0x** [reply]
Thought 1: Drawings would typically reuse the same color many times, so you could simply batch only strokes with the same color. More draw calls, but each draw call is a bit simpler. If you do this, consider using a uniform buffer instead of a instance-mode vertex attribute to pass the color in.

Thought 2: Each instance must have the same size. But you can pad out an instance with degenerate data (like points all at 0,0,0) that won't result in drawing any triangles. You could pick a fixed instance size, which would let you provide colors at a lower rate, at the cost of some wasted cycles on degenerate triangles. I don't know whether this would work well or not! Here's an example with 100 triangles per instance:

- Instance 0: Blue: 100 triangles from stroke 1
- Instance 1: Blue: 20 triangles from stroke 1, 50 triangles from stroke 2, 30 degenerate triangles as padding
- Instance 2: Red: 40 triangles from stroke 3, 60 degenerate triangles as padding

**Reply by kainino0x** [reply]
My intuition is draw calls are really not that expensive, especially if you cache them using a render bundle, and so the first idea would be better.

**Reply by 0x241F31** [reply]
I read another way to provide color is to use **push constants**, so it's unavailable right now.

Here appears another question. 
As I my uniform buffer grows, I need to resize it and recreate all BindGroups, because reference to buffer changes. Is it safe to do that? Will driver automatically utilize unused BindGroups?

Anyway I'm not sure whether this approach better than per-vertex color, as it's required to align uniform groups to 256 bytes, so for every stroke there will be allocated 240 additional bytes of unused space.

**Reply by kainino0x** [reply]
Yes, it's reasonable to recreate bind groups as needed. It's good to reduce the number, but garbage collection should take care of any that you drop.

It's a good point that a uniform buffer will require 256 bytes, I didn't think about that.

**Reply by ghadeeras** [reply]
Assuming I understood the problem :-) ... Maybe you could keep the stroke colors in an unbound buffer (not part of any binding group) that acts like a "palette", and then encode copy commands from it to a small uniform buffer just before draw calls, and only if the stroke color changes. Not sure if that works with batched calls or with render bundles, or if it is efficient, but I suppose you could then keep the buffer terse and maybe reduce a bit the complexity of bind group creation.
